### ISSUE TITLE
Reprovision edge device to different hub.

### ISSUE THREAD

**Issue by denaM314** [Azure/iotedge#6697]
Trying to migrate an initialized edge device from one IoT Hub to another using DPS reprovisioning.
Reprovisioning triggered with a call from postman with the below postman call.

PUT https://global.azure-devices-provisioning.net/{idScope}/registrations/{registrationId}/register?api-version=2021-06-01 

**System Setup:**
- Edge Version: 1.1.15
- DPS Enrollment Group with Symmetric Key attestation.
![image](https://user-images.githubusercontent.com/38193649/193466094-7f18d56d-cc33-4ae7-a5f8-84d1b7fd208f.png)
- Device in question (Hub):
![image](https://user-images.githubusercontent.com/38193649/193466801-f8e298f5-fcc7-4949-b370-816210100746.png)

**Expected result:**
1. DPS Enrollment Group registration record updates to new IoT hub.
2. The destination hub has the device created.
   -  Device data migrated (including existing modules on the device).
   - The edge device is connected to the new hub 
3. The source hub device information has been removed.

**Actual Result:**
1. The above postman call returns that the device is in "Assigning" state. 
2. The DPS Enrollment Group registration record updates correctly and points to new hub.
3.  The destination hub has the device created.
    -  No module data has been migrated.
    -  Destination hub does not show device as connected.
    -   ![image](https://user-images.githubusercontent.com/38193649/193466727-f68e0226-de93-4976-8247-cbd340abad26.png)
4. The source hub still has the device information and showing as connected as if nothing happened.
5.  I do another api call using the Device Registration Status Lookup endpoint to check on the device and it shows as "Assigned"

In summary the DPS portion of the reprovision works as intended but the hub migration does not.

**Comment by jlian** [comment]
A couple questions

1. What do you have, if any, `auto_provisioning_mode` set as in config.toml? 

   ![image](https://user-images.githubusercontent.com/2320572/193647397-eb8591d2-938a-416c-b9da-ded2e2e64979.png)

2. Do you have any IoT edge or device deployments configured in the target IoT hub?

**Comment by denaM314** [comment]
1. If you are referring to the config.yaml file I do not have a "auto_reprovisioning_mode" flag. But I do have a dynamic_reprovisioning flag and that is set to false. I will try to set this to true and see what happens.

![image](https://user-images.githubusercontent.com/38193649/193651769-de716b97-64a5-4c39-83b7-378dfada8646.png)
![image](https://user-images.githubusercontent.com/38193649/193652065-3ad1b45d-5dac-481c-98f2-7e10f388c67e.png)

2. I do not have any deployments configured in the target IoT hub.

**Comment by denaM314** [comment]
> 1. If you are referring to the config.yaml file I do not have a "auto_reprovisioning_mode" flag. But I do have a dynamic_reprovisioning flag and that is set to false. I will try to set this to true and see what happens.
> 
> ![image](https://user-images.githubusercontent.com/38193649/193651769-de716b97-64a5-4c39-83b7-378dfada8646.png) ![image](https://user-images.githubusercontent.com/38193649/193652065-3ad1b45d-5dac-481c-98f2-7e10f388c67e.png)
> 
> 2. I do not have any deployments configured in the target IoT hub.

I ran the reprovisioning with "dynamic_reprovisioning" set to true. 

The same outcome as original post. The only difference is that it looks like edge tried to reprovision with the result below. Was hoping to see some logs but no luck.
![image](https://user-images.githubusercontent.com/38193649/193659980-3e3d4d54-cfab-4f3b-b45f-bc672c3dcdfe.png)

**Comment by jlian** [comment]
Hmm

1. Any chance you could use 1.4 instead or 1.3.1 with EFLOW? 
2. Could you try setting a deployment in the target IoT Hub that matches your original hub?
3. Could you try `iotedge check`?

**Comment by denaM314** [comment]
1. No, we are using windows containers so we have to stick with our current version.
2. The deployment that dictates our company specific modules is built via azure function outside and is called during specific times. **I would assume that the migrate reprovision feature also pushes the running modules as well. Is this correct?**
3. Below is the edge check results.

![image](https://user-images.githubusercontent.com/38193649/193910987-b941bb52-7898-4e86-a890-fef7f20a2882.png)

**Comment by jlian** [comment]
> I would assume that the migrate reprovision feature also pushes the running modules as well. Is this correct?

I don't believe so. Running modules are dictated by the deployment manifest, and the source of truth of that is in IoT Hub, @varunpuranik to keep me honest. Is it possible to have your Azure function also update the deployment in the target hub as well?

Re: provisioning and docker error - so you're not able to see any logs? Could you try some of the steps from here? https://learn.microsoft.com/en-us/azure/iot-edge/troubleshoot?view=iotedge-2018-06#check-the-status-of-the-iot-edge-security-manager-and-its-logs

Also maybe this would help https://github.com/Azure/iotedge/issues/15

**Comment by denaM314** [comment]
OK I was able to get some more information.
1. It seems that the device is stuck in "assigning" state. 
   - This is different from OP because of that setting we changed in the config file.
2. Was able to get some more logs in debug mode (See below)

~~~
10/4/2022 4:42:48 PM debug: tokio_reactor::registration -- scheduling Read for: 0
10/4/2022 4:42:48 PM debug: hyper::proto::h1::io -- flushed 357 bytes
10/4/2022 4:42:48 PM debug: hyper::client::pool -- reuse idle connection for
                     "https://global.azure-devices-provisioning.net"
10/4/2022 4:42:48 PM debug: dps::registration -- Ask DPS for registration status
10/4/2022 4:42:48 PM debug: edgelet_http::client -- Success generating token for request GET 0ne00098D2D/registrations/
                     VM-DM-DEV-4/operations/4.9b0dd2bfa6b3527d.07869e30-cbfb-48ff-8772-6806a970f51f
10/4/2022 4:42:48 PM debug: dps::registration -- Some(DeviceRegistrationResult { tpm: None, x509: None, symmetric_key:
                     None, registration_id: Some("VM-DM-DEV-4"), created_date_time_utc:
                     Some("2022-10-04T21:42:38.311444Z"), assigned_hub: None, device_id: None, status: Some("failed"),
                     substatus: None, error_code: Some(400207), error_message: Some("Custom allocation failed with
                     status code: 500"), last_updated_date_time_utc: Some("2022-10-04T21:42:38.6353445Z"), etag:
                     Some("IjUyMDBlYjhkLTAwMDAtMDcwMC0wMDAwLTYzM2NhOGNlMDAwMCI=") })
10/4/2022 4:42:48 PM warn: edgelet_utils::logging -- Could not provision device
10/4/2022 4:42:48 PM debug: dps::registration -- Device Registration Result: device None, hub None, status
                     Some("failed")
10/4/2022 4:42:48 PM debug: tokio_reactor::registration -- scheduling Read for: 0
10/4/2022 4:42:48 PM warn: edgelet_utils::logging --    caused by: DPS registration failed because the DPS operation is
                     not assigned
10/4/2022 4:42:48 PM error: edgelet_utils::logging -- The daemon could not start up successfully: Could not initialize
                     DPS provisioning client
10/4/2022 4:42:48 PM error: edgelet_utils::logging --   caused by: Could not restore previous provisioning result
10/4/2022 4:42:48 PM debug: tokio_reactor -- dropping I/O source: 0
10/4/2022 4:42:48 PM error: iotedged::windows -- Error while running service. Quitting.
10/4/2022 4:42:48 PM debug: hyper::client::pool -- pooling idle connection for
                     "https://global.azure-devices-provisioning.net"
10/4/2022 4:42:48 PM debug: hyper::proto::h1::conn -- incoming body is chunked encoding
10/4/2022 4:42:48 PM debug: hyper::proto::h1::decode -- incoming chunked header: 0x19E (414 bytes)
10/4/2022 4:42:48 PM debug: hyper::proto::h1::io -- read 671 bytes
10/4/2022 4:42:48 PM debug: hyper::proto::h1::io -- parsed 5 headers
10/4/2022 4:42:48 PM debug: hyper::proto::h1::io -- read 5 bytes
10/4/2022 4:42:48 PM debug: tokio_reactor::registration -- scheduling Read for: 0
10/4/2022 4:42:48 PM debug: tokio_reactor::registration -- scheduling Read for: 0
10/4/2022 4:42:48 PM debug: hyper::proto::h1::conn -- incoming body completed
10/4/2022 4:42:48 PM debug: tokio_reactor::registration -- scheduling Read for: 0
10/4/2022 4:42:48 PM error: edgelet_utils::logging --   caused by: The system cannot find the file specified. (os error
                     2)
10/4/2022 4:42:49 PM info: iotedged -- Starting provisioning edge device via symmetric key...
10/4/2022 4:42:49 PM info: provisioning::provisioning -- Attempting to restore provisioning backup...
10/4/2022 4:42:49 PM info: iotedged -- Provisioning edge device...
10/4/2022 4:42:49 PM info: libiothsm -- (d:\a\_work\1\s\edgelet\hsm-sys\azure-iot-hsm-c\src\hsm_log.c:log_init:41)
                     Initialized logging
10/4/2022 4:42:49 PM info: iotedged -- Finished initializing hsm.
10/4/2022 4:42:49 PM debug: edgelet_http::client -- Success generating token for request PUT
                     0ne00098D2D/registrations/VM-DM-DEV-4/register
10/4/2022 4:42:49 PM debug: hyper::client::connect::dns -- resolving host="global.azure-devices-provisioning.net"
10/4/2022 4:42:49 PM debug: dps::registration -- Obtaining symmetric authentication key
10/4/2022 4:42:49 PM info: provisioning::provisioning -- Provisioning backup could not be restored: Could not restore
                     previous provisioning result
10/4/2022 4:42:49 PM info: dps::registration -- Starting DPS registration with scope_id "0ne00098D2D", registration_id
                     "VM-DM-DEV-4"
10/4/2022 4:42:49 PM info: iotedged::app -- Using config file: C:\ProgramData\iotedge\config.yaml
10/4/2022 4:42:49 PM info: iotedged::windows -- Starting iotedged service.
10/4/2022 4:42:49 PM info: iotedged::app -- Version - 1.1.15 (01685778d8320f126f2daf2f86a78a3d9d4b2f2c)
10/4/2022 4:42:49 PM info: iotedged::windows -- Initializing iotedged service.
10/4/2022 4:42:49 PM info: iotedged::app -- Starting Azure IoT Edge Security Daemon
10/4/2022 4:42:49 PM info: iotedged -- Finished configuring provisioning environment variables and certificates.
10/4/2022 4:42:49 PM info: iotedged -- Initializing hsm...
10/4/2022 4:42:49 PM info: iotedged -- Transparent gateway certificates not found, operating in quick start mode...
10/4/2022 4:42:49 PM info: iotedged -- Configuring C:\ProgramData\iotedge as the home directory.
10/4/2022 4:42:49 PM info: iotedged -- Configuring certificates...
10/4/2022 4:42:49 PM debug: tokio_reactor::registration -- scheduling Write for: 0
10/4/2022 4:42:49 PM debug: tokio_reactor -- adding I/O source: 0
10/4/2022 4:42:49 PM debug: hyper::client::connect::http -- connecting to 23.96.222.45:443
10/4/2022 4:42:49 PM debug: tokio_reactor::registration -- scheduling Read for: 0
10/4/2022 4:42:49 PM debug: tokio_reactor::registration -- scheduling Read for: 0
10/4/2022 4:42:49 PM debug: hyper::client::connect::http -- connected to Some(23.96.222.45:443)
10/4/2022 4:42:49 PM debug: tokio_reactor::registration -- scheduling Read for: 0
10/4/2022 4:42:49 PM debug: hyper::proto::h1::io -- flushed 355 bytes
10/4/2022 4:42:49 PM debug: hyper::proto::h1::io -- flushed 32 bytes
10/4/2022 4:42:49 PM debug: tokio_reactor::registration -- scheduling Read for: 0
10/4/2022 4:42:49 PM debug: tokio_reactor::registration -- scheduling Read for: 0
10/4/2022 4:42:49 PM debug: hyper::proto::h1::conn -- incoming body is chunked encoding
10/4/2022 4:42:49 PM debug: hyper::proto::h1::decode -- incoming chunked header: 0x5E (94 bytes)
10/4/2022 4:42:49 PM debug: hyper::proto::h1::io -- read 476 bytes
10/4/2022 4:42:49 PM debug: hyper::proto::h1::io -- parsed 7 headers
10/4/2022 4:42:49 PM debug: dps::registration -- Ask DPS for registration status
10/4/2022 4:42:49 PM debug: dps::registration -- DPS registration result will retry 13 times every 10 seconds
10/4/2022 4:42:49 PM debug: tokio_reactor::registration -- scheduling Read for: 0
10/4/2022 4:42:49 PM debug: edgelet_http::client -- Success generating token for request GET 0ne00098D2D/registrations/
                     VM-DM-DEV-4/operations/4.9b0dd2bfa6b3527d.ddbb2fbc-daed-4917-86f8-76842e19c4c0
10/4/2022 4:42:49 PM debug: hyper::proto::h1::io -- flushed 361 bytes
10/4/2022 4:42:49 PM debug: tokio_reactor::registration -- scheduling Read for: 0
10/4/2022 4:42:49 PM debug: hyper::client::pool -- reuse idle connection for
                     "https://global.azure-devices-provisioning.net"
10/4/2022 4:42:49 PM debug: tokio_reactor::registration -- scheduling Read for: 0
10/4/2022 4:42:49 PM debug: hyper::proto::h1::conn -- incoming body completed
10/4/2022 4:42:49 PM debug: hyper::proto::h1::io -- read 5 bytes
10/4/2022 4:42:49 PM debug: tokio_reactor::registration -- scheduling Read for: 0
10/4/2022 4:42:49 PM debug: dps::registration -- Some(RegistrationOperationStatus { operation_id:
                     "4.9b0dd2bfa6b3527d.ddbb2fbc-daed-4917-86f8-76842e19c4c0", status: Some("assigning"),
                     registration_state: None })
10/4/2022 4:42:49 PM debug: hyper::client::pool -- pooling idle connection for
                     "https://global.azure-devices-provisioning.net"
10/4/2022 4:42:49 PM debug: tokio_reactor::registration -- scheduling Read for: 0
10/4/2022 4:42:49 PM debug: tokio_reactor::registration -- scheduling Read for: 0
10/4/2022 4:42:49 PM debug: tokio_reactor::registration -- scheduling Read for: 0
10/4/2022 4:42:49 PM debug: tokio_reactor::registration -- scheduling Read for: 0
10/4/2022 4:42:49 PM debug: tokio_reactor::registration -- scheduling Read for: 0
10/4/2022 4:42:49 PM debug: dps::registration -- Device Registration Result: device None, hub None, status
                     Some("assigning")
10/4/2022 4:42:49 PM debug: hyper::client::pool -- pooling idle connection for
                     "https://global.azure-devices-provisioning.net"
10/4/2022 4:42:49 PM debug: hyper::proto::h1::conn -- incoming body is chunked encoding
10/4/2022 4:42:49 PM debug: hyper::proto::h1::io -- parsed 7 headers
10/4/2022 4:42:49 PM debug: hyper::proto::h1::io -- read 608 bytes
10/4/2022 4:42:49 PM debug: hyper::proto::h1::conn -- incoming body completed
10/4/2022 4:42:49 PM debug: hyper::proto::h1::io -- read 5 bytes
10/4/2022 4:42:49 PM debug: hyper::proto::h1::decode -- incoming chunked header: 0xA8 (168 bytes)
~~~

**Comment by jlian** [comment]
Let me check internally with some DPS experts

**Comment by denaM314** [comment]
Api Request from postman
![image](https://user-images.githubusercontent.com/38193649/194409513-1274438b-aec9-4876-90b8-d9e70a855ab0.png)

New registration record (It is correct):
![image](https://user-images.githubusercontent.com/38193649/194410601-b5c7ca5e-e0e1-436b-89bb-5fd0596d810f.png)

**Comment by khilscher** [comment]
I was looking for the request/response between DPS and your Azure Function, not the device to DPS.

But I am also seeing you are trying to pass in a desired Hub in the custom payload between the device and DPS. Why? This is counter to what DPS is supposed to do. You link DPS to N Hubs, then assign an allocation policy in the DPS enrollment that makes the decision as to which Hub to assign. DPS then registers that device in Hub and returns the Hub endpoint back to the device. The device should never have to pass in a desired Hub.

**Comment by denaM314** [comment]
I am using custom allocation in this test environment since I am testing reprovisioning and migrating from one hub to another. The payload allows the http trigger to request reprovision to the same hub every time. This gave me repeatability in testing.

Basically I initialize a device in hub "A" and make sure its working properly then I try to reprovision the device to hub "B". This is not a production setup.

Below is the Azure function logs that handle the custom allocation to DPS.
![image](https://user-images.githubusercontent.com/38193649/194426832-9b74d33e-9bf0-473b-a62a-ea2eda6eaa56.png)

**Comment by khilscher** [comment]
Thanks. To summarize:

1. You are reprovisioning the same Hub, correct? 
2. Were you also trying to reprovision to the same Hub above when you created this issue? I'm assuming not, and that you are actually trying to have your Edge device re-provision to another hub right?
3. Your enrollment is set to "Re-provision and migrate data" correct?
4. DPS returns "assigning" and remains stuck in the "assigning" state correct?
5. Can you send us the latest operationId where the state is stuck in "assigning"?

**Comment by denaM314** [comment]
1. I always use the same two hubs. Provision to Hub A then reprovision to Hub B.
2. Correct, a different Hub.
3. Correct
4. No, If I recheck the registration status using the api, it comes back as assigned.
   - ~~~
      {
        "registrationId": "VM-DM-DEV-4",
        "createdDateTimeUtc": "2022-10-06T22:02:23.8887596Z",
        "assignedHub": "test-hub-integration.azure-devices.net",
        "deviceId": "VM-DM-DEV-4",
        "status": "assigned",
        "substatus": "deviceDataMigrated",
        "lastUpdatedDateTimeUtc": "2022-10-06T22:02:24.5789073Z",
        "etag": "IjZkMDE3ZWM1LTAwMDAtMDcwMC0wMDAwLTYzM2Y1MDcwMDAwMCI="
     }
     ~~~
5. Latest OperId: 4.9b0dd2bfa6b3527d.c041b90c-39b3-4610-9435-b7710db951ee


Currently the devices gets removed from the original Hub and gets created in the target Hub. The target Hub does not show that it is  connected to it. (See Below)

Target Hub:
![image](https://user-images.githubusercontent.com/38193649/194555207-50b027a7-033c-42e8-81d8-7f5bd730dabb.png)

**Comment by jlian** [comment]
@khilscher does this look like a DPS issue to you or IoT Edge?

**Comment by khilscher** [comment]
@jlian I can't really tell and I don't have access to the Kusto logs to verify using the operationID. Suggest @denaM314 open a support case via the Azure Portal so we can have a deeper investigation.

**Comment by raisalitch** [comment]
@denaM314 Did you end up opening a support case for this?

**Comment by raisalitch** [comment]
Hey @denaM314 I'm going to close this but please open up a support case for this if you need further help. Thanks!
